docs/design/verification_protocol.md

```diff
@@ -0,0 +1,122 @@
+
+
+## High Level Flow
+
+1. App user is diagnosed with Covid-19. PHA issues a PIN code to the user.
+2. The user enters the PIN code on their App
+3. The app creates a HMAC value that combines the TEKs from the device along
+   with calculated transmission risk value, rolling period start and rolling
+	 period count values (the key data + metadata).
+	 These values are sent to the PHA server that issued the PIN.
+4. If the PIN is valid, the PHA issues a JWT that is signed using ECDSA over
+   the P-256 elliptic curve with SHA-256 as a hash function. The JWT includes
+   additional claims about the data (see below).
+5. The app on the user's device sends this signed JWT to the exposure
+   notifications server (this project).
+6. If the JWT is valid and signed by a trusted PHA (verified because the public
+	 key has been previously shared with the server), then the keys are imported
+	 into the server for distribution to other devices in the geography. Because
+   we are using pre-shared public keys, this verification can be done with
+   offline verification.
+
+## JWT Verification + Accepted Claims
+
+The exposure notifications server will accept this JWT in the current
+`verificationPayload` field in the exposure reporting API. When this is present,
+device attestations should be disabled. In addition, the exposure publishing
+API will also accept a new field `hmackey` to be used as the key for
+re-calculating the HMAC that was signed by the PHA verification server. The
+`hmackey` field must be base64 encoded.
+
+The JWT itself will be unpacked and the following validations will be performed.
+First, using the standard claims.
+
+* `iss` : The issuer will be used to determine which public key(s) are valid for
+verification. This is to allow for key rotation.
+* `aud` : The audience must be as configured for this installation of the
+exposure notifications server.
+* `iat` : The unix timestamp at which the token was issued.
+* `exp` : The unix timestamp at which the token will expire.
+* `nbf` : If present, the "not before" timestamp will be honored.
+
+We also prescribe a set of private claims to transmit data from the PHA
+verification server to the exposure notifications server.
+
+* `tekhmac` : The HMAC of the TEKs that was presented to the PHA verification
+server. This must be calculated in a specific way (see below).
+* `phadata` : Contains a map of claims for the PHA to communicate back
+to the mobile app and/or the exposure notifications server. This specific
+implementation of the exposure notifications server will disregard and never
+save the `phadata`
+* `trisk` : Contains an array of transmission risk overrides to enact when
+importing the associated keys. If data is present in this field, it will
+override the data in the upload from the device.
+* `keyVersion` : The version of the public key to use for verification.
+
+### Transmission Risk Overrides
+
+An array of JSON objects of the following structure:
```

Comment:
would it be useful to require these be ordered? or should we specify sorting on the `sinceRollingPeriod` before computation

Reply:
we can suggest it - the server will sort upon receipt either way